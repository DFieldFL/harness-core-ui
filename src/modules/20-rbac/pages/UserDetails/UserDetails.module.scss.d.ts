/* eslint-disable */
// this is an auto-generated file
declare const styles: {
  readonly header: string
  readonly body: string
  readonly card: string
  readonly addButton: string
}
export default styles
