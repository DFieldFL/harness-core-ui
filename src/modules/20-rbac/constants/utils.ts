export enum RbacResourceGroupTypes {
  DYNAMIC_RESOURCE_SELECTOR = 'DynamicResourceSelector'
}
