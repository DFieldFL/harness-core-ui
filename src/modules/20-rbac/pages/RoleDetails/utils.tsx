import produce from 'immer'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import type { ResourceType } from '@common/interfaces/ResourceInterface'
import type { Permission, PermissionResponse } from 'services/rbac'

export const getPermissionMap = (permissionList?: PermissionResponse[]): Map<ResourceType, Permission[]> => {
  const permissionsMap: Map<ResourceType, Permission[]> = new Map()
  permissionList?.map(response => {
    if (response.permission.resourceType) {
      if (permissionsMap.has(response.permission.resourceType as ResourceType)) {
        permissionsMap.get(response.permission.resourceType as ResourceType)?.push(response.permission)
      } else permissionsMap.set(response.permission.resourceType as ResourceType, [response.permission])
    }
  })
  return permissionsMap
}

export const onPermissionChange = (
  permission: string,
  isAdd: boolean,
  permissions: string[],
  setPermissions: (value: React.SetStateAction<string[]>) => void
): void => {
  if (isAdd) {
    if (
      permission === PermissionIdentifier.EDIT_DASHBOARD ||
      permissions.indexOf(PermissionIdentifier.EDIT_DASHBOARD) !== -1
    ) {
      setPermissions(_permissions => [...permissions, permission, PermissionIdentifier.VIEW_DASHBOARD])
    } else {
      setPermissions(_permissions => [...permissions, permission])
    }
  } else if (
    !(
      permission === PermissionIdentifier.VIEW_DASHBOARD &&
      permissions.indexOf(PermissionIdentifier.EDIT_DASHBOARD) !== -1
    )
  ) {
    setPermissions(_permissions =>
      produce(_permissions, draft => {
        draft?.splice(permissions.indexOf(permission), 1)
      })
    )
  }
}
