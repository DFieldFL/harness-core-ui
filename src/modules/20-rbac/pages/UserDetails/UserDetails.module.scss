.header {
  align-items: flex-start !important;
  padding-top: var(--spacing-large) !important;
}

.body {
  padding: var(--spacing-large) !important;
  display: flex;
  flex-direction: row;
}

.card {
  box-shadow: var(--card-shadow);
  align-items: center;
  display: flex !important;
  flex-wrap: wrap;
  gap: var(--spacing-xsmall);
}

.name {
  overflow-wrap: anywhere;
}
.lockIcon {
  border-radius: 100% !important;
}
.popover {
  display: block;
  max-width: 250px;
  padding: var(--spacing-small);
}
