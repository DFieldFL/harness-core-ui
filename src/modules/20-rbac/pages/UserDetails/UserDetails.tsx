import React from 'react'
import { Text, Layout, Color, Card, Avatar } from '@wings-software/uicore'
import { useParams } from 'react-router-dom'
import { useStrings } from 'framework/strings'
import { useGetAggregatedUser } from 'services/cd-ng'
import { Page } from '@common/exports'
import routes from '@common/RouteDefinitions'
import { Breadcrumbs } from '@common/components/Breadcrumbs/Breadcrumbs'
import { PageSpinner } from '@common/components'
import { PageError } from '@common/components/Page/PageError'
import RoleBindingsList from '@rbac/components/RoleBindingsList/RoleBindingsList'
import type { PipelineType, ProjectPathProps, UserPathProps } from '@common/interfaces/RouteInterfaces'
import { PrincipalType, useRoleAssignmentModal } from '@rbac/modals/RoleAssignmentModal/useRoleAssignmentModal'
import { useDocumentTitle } from '@common/hooks/useDocumentTitle'
import { ResourceType } from '@rbac/interfaces/ResourceType'
import ManagePrincipalButton from '@rbac/components/ManagePrincipalButton/ManagePrincipalButton'
import UserGroupTable from './views/UserGroupTable'
import css from './UserDetails.module.scss'

const UserDetails: React.FC = () => {
  const { getString } = useStrings()
  const { accountId, orgIdentifier, projectIdentifier, module, userIdentifier } = useParams<
    PipelineType<ProjectPathProps & UserPathProps>
  >()

  const { data, loading, error, refetch } = useGetAggregatedUser({
    userId: userIdentifier,
    queryParams: {
      accountIdentifier: accountId,
      orgIdentifier,
      projectIdentifier
    }
  })

  const { openRoleAssignmentModal } = useRoleAssignmentModal({
    onSuccess: refetch
  })

  const user = data?.data?.user

  useDocumentTitle([user?.name || '', getString('users')])

  if (loading) return <PageSpinner />
  if (error) return <PageError message={error.message} onClick={() => refetch()} />
  if (!user) return <></>
  return (
    <>
      <Page.Header
        size="xlarge"
        className={css.header}
        title={
          <Layout.Vertical>
            <Breadcrumbs
              links={[
                {
                  url: routes.toAccessControl({ accountId, orgIdentifier, projectIdentifier, module }),
                  label: getString('accessControl')
                },
                {
                  url: routes.toUsers({ accountId, orgIdentifier, projectIdentifier, module }),
                  label: getString('users')
                },
                {
                  url: '#',
                  label: user.name || user.email
                }
              ]}
            />
            <Layout.Horizontal flex={{ alignItems: 'center', justifyContent: 'flex-start' }} spacing="medium">
              <Avatar name={user.name} email={user.email} size="large" hoverCard={false} />
              <Layout.Vertical padding={{ left: 'medium' }} spacing="xsmall">
                <Text color={Color.BLACK} font="medium">
                  {user.name}
                </Text>
                <Text>{user.email}</Text>
              </Layout.Vertical>
            </Layout.Horizontal>
          </Layout.Vertical>
        }
      />
      <Page.Body className={css.body}>
        <Layout.Vertical width="100%" padding="large">
          <UserGroupTable />
          <Layout.Vertical spacing="medium" padding={{ bottom: 'large' }}>
            <Text color={Color.BLACK} font={{ size: 'medium', weight: 'semi-bold' }}>
              {getString('rbac.roleBinding')}
            </Text>
            <Card className={css.card}>
              <RoleBindingsList data={data?.data?.roleAssignmentMetadata} />
            </Card>
            <Layout.Horizontal
              flex={{ alignItems: 'center', justifyContent: 'flex-start' }}
              padding={{ top: 'medium' }}
            >
              <ManagePrincipalButton
                data-testid={'addRole-UserGroup'}
                text={getString('common.plusNumber', { number: getString('common.role') })}
                minimal
                intent="primary"
                onClick={event => {
                  event.stopPropagation()
                  openRoleAssignmentModal(PrincipalType.USER, user, data?.data?.roleAssignmentMetadata)
                }}
                resourceIdentifier={user.uuid}
                resourceType={ResourceType.USER}
              />
            </Layout.Horizontal>
          </Layout.Vertical>
        </Layout.Vertical>
      </Page.Body>
    </>
  )
}

export default UserDetails
