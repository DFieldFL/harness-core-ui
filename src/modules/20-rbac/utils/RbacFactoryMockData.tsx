import React from 'react'
import type { ResourceHandler, ResourceTypeGroupHandler } from '@rbac/factories/RbacFactory'
import { ResourceType, ResourceTypeGroup } from '@rbac/interfaces/ResourceType'

export const getResourceTypeHandlerMock = (resource: ResourceType): ResourceHandler | undefined => {
  switch (resource) {
    case ResourceType.PROJECT:
      return {
        icon: 'nav-project',
        label: 'Projects',
        // eslint-disable-next-line react/display-name
        addResourceModalBody: () => <></>
      }
    case ResourceType.ORGANIZATION:
      return {
        icon: 'nav-project',
        label: 'Organizations',
        // eslint-disable-next-line react/display-name
        addResourceModalBody: () => <></>
      }
    case ResourceType.SECRET:
      return {
        icon: 'lock',
        label: 'Secrets',
        category: ResourceTypeGroup.PROJECT_RESOURCES,
        // eslint-disable-next-line react/display-name
        addResourceModalBody: () => <></>
      }
  }
}

export const getResourceGroupTypeHandlerMock = (
  resource: ResourceType | ResourceTypeGroup
): ResourceTypeGroupHandler | undefined => {
  switch (resource) {
    case ResourceTypeGroup.PROJECT_RESOURCES:
      return {
        icon: 'nav-project',
        label: 'Project Resources',
        resourceTypes: new Set([ResourceType.SECRET])
      }
    case ResourceTypeGroup.ADMINSTRATIVE_FUNCTIONS:
      return {
        icon: 'nav-project',
        label: 'Administrative Fucntions',
        resourceTypes: new Set([ResourceType.ORGANIZATION, ResourceType.PROJECT])
      }
  }
}
