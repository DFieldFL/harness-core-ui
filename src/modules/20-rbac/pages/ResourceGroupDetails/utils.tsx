import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import type { ResourceSelector } from 'services/cd-ng'

export const getSelectedResourcesMap = (
  resourceSelectorList?: ResourceSelector[]
): Map<ResourceType, string[] | string> => {
  const map: Map<ResourceType, string[] | string> = new Map()

  resourceSelectorList?.map(resourceSelector => {
    if (resourceSelector.type === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      map.set(resourceSelector.resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    else map.set(resourceSelector.resourceType, resourceSelector.identifiers)
  })

  return map
}

export const getResourceSelectorsfromMap = (map: Map<ResourceType, string[] | string>): ResourceSelector[] => {
  const resourceSelectors: ResourceSelector[] = []
  map.forEach((value, key) => {
    if (value === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      resourceSelectors.push({
        type: RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR,
        resourceType: key
      })
    else {
      resourceSelectors.push({
        type: RbacResourceGroupTypes.STATIC_RESOURCE_SELECTOR,
        resourceType: key,
        identifiers: value
      })
    }
  })
  return resourceSelectors
}
