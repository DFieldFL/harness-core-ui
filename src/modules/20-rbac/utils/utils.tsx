import React, { ReactNode } from 'react'
import { Menu } from '@blueprintjs/core'
import type { ItemRenderer } from '@blueprintjs/select'
import {
  IconName,
  Layout,
  MultiSelectOption,
  Text,
  Avatar,
  Color,
  ModalErrorHandlerBinding,
  getErrorInfoFromErrorObject,
  SelectOption
} from '@wings-software/uicore'
import { defaultTo } from 'lodash-es'
import type { StringsMap } from 'stringTypes'
import type { AccessControlCheckError, RoleAssignmentMetadataDTO, UserMetadataDTO } from 'services/cd-ng'
import { Scope } from '@common/interfaces/SecretsInterface'
import type {
  Assignment,
  RoleOption,
  ResourceGroupOption
} from '@rbac/modals/RoleAssignmentModal/views/UserRoleAssigment'
import { isEmail } from '@common/utils/Validation'
import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import RBACTooltip from '@rbac/components/RBACTooltip/RBACTooltip'
import type { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import type { ResourceType } from '@common/interfaces/ResourceInterface'
import type { FeatureRequest } from 'framework/featureStore/featureStoreUtil'
import type { PermissionsRequest } from '@rbac/hooks/usePermission'
import { FeatureWarningTooltip } from '@common/components/FeatureWarning/FeatureWarningWithTooltip'
import type { UseStringsReturn } from 'framework/strings'
import css from './utils.module.scss'

export interface UserItem extends MultiSelectOption {
  email?: string
}

export enum PrincipalType {
  USER = 'USER',
  USER_GROUP = 'USER_GROUP',
  SERVICE = 'SERVICE_ACCOUNT'
}

export const getRoleIcon = (roleIdentifier: string): IconName => {
  switch (roleIdentifier) {
    case '_account_viewer':
    case '_organization_viewer':
    case '_project_viewer':
      return 'viewerRole'
    case '_account_admin':
    case '_organization_admin':
    case '_project_admin':
      return 'adminRole'
    case '_pipeline_executor':
      return 'pipeline-executor'
    default:
      return 'customRole'
  }
}

export const UserTagRenderer = (item: UserItem, validate = false): React.ReactNode => (
  <Layout.Horizontal key={item.value.toString()} flex spacing="small">
    <Avatar name={item.label} email={item.value.toString()} size="xsmall" hoverCard={false} />
    <Text color={validate && !isEmail(item.value.toString().toLowerCase()) ? Color.RED_500 : Color.BLACK}>
      {item.label}
    </Text>
  </Layout.Horizontal>
)

export const UserItemRenderer: ItemRenderer<UserItem> = (item, { handleClick }) => (
  <Menu.Item
    key={item.value.toString()}
    text={
      <Layout.Horizontal spacing="small" flex={{ alignItems: 'center', justifyContent: 'flex-start' }}>
        <Avatar name={item.label} email={item.email || item.value.toString()} size="small" hoverCard={false} />
        <Layout.Vertical padding={{ left: 'small' }}>
          <Text color={Color.BLACK}>{item.label}</Text>
          <Text color={Color.GREY_700}>{item.email || item.value}</Text>
        </Layout.Vertical>
      </Layout.Horizontal>
    }
    onClick={handleClick}
  />
)

export enum InvitationStatus {
  USER_INVITED_SUCCESSFULLY = 'USER_INVITED_SUCCESSFULLY',
  USER_ADDED_SUCCESSFULLY = 'USER_ADDED_SUCCESSFULLY',
  USER_ALREADY_ADDED = 'USER_ALREADY_ADDED',
  USER_ALREADY_INVITED = 'USER_ALREADY_INVITED',
  FAIL = 'FAIL'
}

interface HandleInvitationResponse {
  responseType: InvitationStatus
  getString: (key: keyof StringsMap, vars?: Record<string, any> | undefined) => string
  showSuccess: (message: string | ReactNode, timeout?: number, key?: string) => void
  modalErrorHandler?: ModalErrorHandlerBinding
  onSubmit?: () => void
  onUserAdded?: () => void
}

export const handleInvitationResponse = ({
  responseType,
  getString,
  showSuccess,
  modalErrorHandler,
  onSubmit,
  onUserAdded
}: HandleInvitationResponse): void => {
  switch (responseType) {
    case InvitationStatus.USER_INVITED_SUCCESSFULLY: {
      onSubmit?.()
      return showSuccess(getString('rbac.usersPage.invitationSuccess'))
    }
    case InvitationStatus.USER_ADDED_SUCCESSFULLY: {
      onUserAdded?.()
      return showSuccess(getString('rbac.usersPage.userAddedSuccess'))
    }
    case InvitationStatus.USER_ALREADY_ADDED:
      return showSuccess(getString('rbac.usersPage.userAlreadyAdded'))
    case InvitationStatus.USER_ALREADY_INVITED:
      return showSuccess(getString('rbac.usersPage.userAlreadyInvited'))
    default:
      return modalErrorHandler?.showDanger(getString('rbac.usersPage.invitationError'))
  }
}

export const getScopeBasedManagedResourceGroup = (
  scope: Scope,
  getString: UseStringsReturn['getString']
): SelectOption => {
  switch (scope) {
    case Scope.ACCOUNT:
      return {
        label: getString('rbac.allAccountResources'),
        value: '_all_account_level_resources'
      }
    case Scope.ORG:
      return {
        label: getString('rbac.allOrgResources'),
        value: '_all_organization_level_resources'
      }
    case Scope.PROJECT:
      return {
        label: getString('rbac.allProjectResources'),
        value: '_all_project_level_resources'
      }
    default:
      return {
        label: getString('rbac.allResources'),
        value: '_all_resources'
      }
  }
}

export const getScopeBasedDefaultAssignment = (
  scope: Scope,
  getString: UseStringsReturn['getString'],
  isCommunity: boolean
): Assignment[] => {
  if (isCommunity) {
    return []
  } else {
    const resourceGroup: ResourceGroupOption = {
      managedRoleAssignment: true,
      ...getScopeBasedManagedResourceGroup(scope, getString)
    }
    switch (scope) {
      case Scope.ACCOUNT:
        return [
          {
            role: {
              label: getString('common.accViewer'),
              value: '_account_viewer',
              managed: true,
              managedRoleAssignment: true
            },
            resourceGroup
          }
        ]
      case Scope.ORG:
        return [
          {
            role: {
              label: getString('common.orgViewer'),
              value: '_organization_viewer',
              managed: true,
              managedRoleAssignment: true
            },
            resourceGroup
          }
        ]
      case Scope.PROJECT:
        return [
          {
            role: {
              label: getString('common.projectViewer'),
              value: '_project_viewer',
              managed: true,
              managedRoleAssignment: true
            },
            resourceGroup
          }
        ]
      default:
        return []
    }
  }
}

export const isAssignmentFieldDisabled = (value: RoleOption | ResourceGroupOption): boolean => {
  if (value.assignmentIdentifier || value.managedRoleAssignment) {
    return true
  }
  return false
}
export const isDynamicResourceSelector = (value: string | string[]): boolean => {
  if (value === RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR) {
    return true
  }
  return false
}

interface ErrorHandlerProps {
  data: AccessControlCheckError
}

export const getRBACErrorMessage = (error: ErrorHandlerProps): string | React.ReactElement => {
  const err = error?.data
  if (err?.code === 'NG_ACCESS_DENIED' && err?.failedPermissionChecks?.length) {
    const { permission, resourceType, resourceScope } = err.failedPermissionChecks[0]
    if (permission && resourceType && resourceScope) {
      return (
        <RBACTooltip
          permission={permission as PermissionIdentifier}
          resourceType={resourceType as ResourceType}
          resourceScope={resourceScope}
          className={css.tooltip}
        />
      )
    }
  }
  return getErrorInfoFromErrorObject(error)
}

export const getAssignments = (roleBindings: RoleAssignmentMetadataDTO[]): Assignment[] => {
  return (
    roleBindings?.map(roleAssignment => {
      return {
        role: {
          label: roleAssignment.roleName,
          value: roleAssignment.roleIdentifier,
          managed: roleAssignment.managedRole,
          assignmentIdentifier: roleAssignment.identifier,
          managedRoleAssignment: roleAssignment.managedRoleAssignment
        },
        resourceGroup: {
          label: roleAssignment.resourceGroupName || '',
          value: roleAssignment.resourceGroupIdentifier || '',
          managedRoleAssignment: roleAssignment.managedRoleAssignment,
          assignmentIdentifier: roleAssignment.identifier
        }
      }
    }) || []
  )
}

interface FeatureProps {
  featureRequest: FeatureRequest
  isPermissionPrioritized?: boolean
}

interface TooltipProps {
  permissionRequest?: Omit<PermissionsRequest, 'permissions'> & { permission: PermissionIdentifier }
  featureProps?: FeatureProps
  canDoAction: boolean
  featureEnabled: boolean
}

interface TooltipReturn {
  tooltip?: React.ReactElement
}

export function getTooltip({
  permissionRequest,
  featureProps,
  canDoAction,
  featureEnabled
}: TooltipProps): TooltipReturn {
  // if permission check override the priorirty
  if (featureProps?.isPermissionPrioritized && permissionRequest && !canDoAction) {
    return {
      tooltip: (
        <RBACTooltip
          permission={permissionRequest.permission}
          resourceType={permissionRequest.resource.resourceType}
          resourceScope={permissionRequest.resourceScope}
        />
      )
    }
  }

  // feature check by default take priority
  if (featureProps?.featureRequest && !featureEnabled) {
    return {
      tooltip: <FeatureWarningTooltip featureName={featureProps?.featureRequest.featureName} />
    }
  }

  // permission check
  if (permissionRequest && !canDoAction) {
    return {
      tooltip: (
        <RBACTooltip
          permission={permissionRequest.permission}
          resourceType={permissionRequest.resource.resourceType}
          resourceScope={permissionRequest.resourceScope}
        />
      )
    }
  }

  return {}
}

export const getUserName = (user: UserMetadataDTO): string => {
  return defaultTo(user.name, user.email)
}
