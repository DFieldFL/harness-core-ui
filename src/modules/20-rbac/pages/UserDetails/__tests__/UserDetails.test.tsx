import React from 'react'
import { render, RenderResult, waitFor } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import routes from '@common/RouteDefinitions'
import { accountPathProps, userPathProps } from '@common/utils/routeUtils'
import UserDetails from '../UserDetails'
import { userInfo } from './mock'

jest.mock('services/cd-ng', () => ({
  useGetAggregatedUser: jest.fn().mockImplementation(() => {
    return { data: userInfo, refetch: jest.fn(), error: null, loading: false }
  })
}))

jest.mock('react-timeago', () => () => 'dummy date')

describe('UserDetails Test', () => {
  let container: HTMLElement
  let getAllByText: RenderResult['getAllByText']

  beforeEach(async () => {
    const renderObj = render(
      <TestWrapper
        path={routes.toUserDetails({ ...accountPathProps, ...userPathProps })}
        pathParams={{ accountId: 'testAcc', userIdentifier: 'dummy' }}
      >
        <UserDetails />
      </TestWrapper>
    )
    container = renderObj.container
    getAllByText = renderObj.getAllByText
    await waitFor(() => getAllByText('accessControl'))
  })
  test('render data', () => {
    expect(container).toMatchSnapshot()
  })
})
