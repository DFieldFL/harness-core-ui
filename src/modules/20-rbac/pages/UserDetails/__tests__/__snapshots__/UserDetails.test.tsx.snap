// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails Test render data 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main container xlarge header StyledProps--flex StyledProps--padding-left-large StyledProps--padding-right-large"
  >
    <div
      class="Layout--vertical StyledProps--main"
    >
      <div
        class="breadcrumbs"
      >
        <div>
          <a
            class="breadcrumb"
            href="/account/testAcc/home/setup/access-control"
          >
            accessControl
          </a>
          <span
            class="bp3-icon StyledProps--main separator"
            data-icon="main-chevron-right"
          >
            <svg
              height="11"
              viewBox="0 0 8 14"
              width="11"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.7 6.3l-6-6C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7L5.6 7 .3 12.3c-.2.2-.3.4-.3.7 0 .6.4 1 1 1 .3 0 .5-.1.7-.3l6-6c.2-.2.3-.4.3-.7s-.1-.5-.3-.7z"
                fill="#4d5969"
              />
            </svg>
          </span>
        </div>
        <div>
          <a
            class="breadcrumb"
            href="/account/testAcc/home/setup/access-control/users"
          >
            users
          </a>
          <span
            class="bp3-icon StyledProps--main separator"
            data-icon="main-chevron-right"
          >
            <svg
              height="11"
              viewBox="0 0 8 14"
              width="11"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.7 6.3l-6-6C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7L5.6 7 .3 12.3c-.2.2-.3.4-.3.7 0 .6.4 1 1 1 .3 0 .5-.1.7-.3l6-6c.2-.2.3-.4.3-.7s-.1-.5-.3-.7z"
                fill="#4d5969"
              />
            </svg>
          </span>
        </div>
        <div>
          <span
            class="lastBreadcrumb"
          >
            dummy
          </span>
        </div>
      </div>
      <div
        class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-flex-start"
      >
        <div
          class="Avatar--Avatar Avatar--contentStyle"
        >
          <span
            class="bp3-popover-wrapper Avatar--avatarPopOver"
          >
            <span
              class="bp3-popover-target"
            >
              <div
                class="Avatar--AvatarInner"
                style="border-radius: 100%; line-height: 66px; width: 72px; height: 72px; font-size: 28px;"
                tabindex="0"
              >
                D
              </div>
            </span>
          </span>
        </div>
        <div
          class="Layout--vertical Layout--layout-spacing-xsmall StyledProps--main StyledProps--padding-left-medium"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-black StyledProps--font-medium"
          >
            dummy
          </p>
          <p
            class="StyledProps--font StyledProps--main"
          >
            dummy@harness.io
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="StyledProps--font StyledProps--main pageBody body"
    data-name="page-body"
  >
    <div
      class="Layout--vertical StyledProps--main StyledProps--padding-large"
      style="width: 100%;"
    >
      <div>
        <p
          class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-black StyledProps--font-size-medium StyledProps--font-weight-semi-bold StyledProps--padding-bottom-medium"
        >
          common.userGroups
        </p>
        <div
          class="table"
        >
          <div
            class="body"
          >
            <div
              class="row card"
              role="row"
            >
              <div
                class="cell"
                role="cell"
                style="width: 95%;"
              >
                <div
                  class="Layout--horizontal Layout--layout-spacing-small StyledProps--main name"
                >
                  <p
                    class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-black"
                    style="--text-line-clamp: 1;"
                  >
                    testGroup
                  </p>
                </div>
              </div>
              <div
                class="cell"
                role="cell"
                style="width: 5%;"
              >
                <span
                  class="bp3-popover-wrapper bp3-dark"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <button
                      class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow"
                      data-testid="menu-UserGroup-testGroup"
                      type="button"
                    >
                      <span
                        class="bp3-icon StyledProps--main"
                        data-icon="Options"
                      >
                        <svg
                          fill="none"
                          height="16"
                          viewBox="0 0 4 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            clip-rule="evenodd"
                            d="M2 12.5a1 1 0 100 2 1 1 0 000-2zm-2 1a2 2 0 114 0 2 2 0 01-4 0zm2-12a1 1 0 100 2 1 1 0 000-2zm-2 1a2 2 0 114 0 2 2 0 01-4 0zM2 7a1 1 0 100 2 1 1 0 000-2zM0 8a2 2 0 114 0 2 2 0 01-4 0z"
                            fill="currentColor"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </span>
                    </button>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="Layout--vertical Layout--layout-spacing-medium StyledProps--main StyledProps--padding-bottom-large"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-black StyledProps--font-size-medium StyledProps--font-weight-semi-bold"
        >
          rbac.roleBinding
        </p>
        <div
          class="bp3-card bp3-elevation-0 Card--card card"
        >
          <span
            class="bp3-tag customtag"
          >
            <span
              class="bp3-text-overflow-ellipsis bp3-fill"
            >
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/roles/New_Role"
              >
                New Role
              </a>
               - 
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/resource-groups/_all_resources"
              >
                All Resources
              </a>
            </span>
          </span>
          <span
            class="bp3-tag customtag"
          >
            <span
              class="bp3-text-overflow-ellipsis bp3-fill"
            >
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/roles/TestRole"
              >
                TestRole
              </a>
               - 
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/resource-groups/_all_resources"
              >
                All Resources
              </a>
            </span>
          </span>
          <span
            class="bp3-tag harnesstag"
          >
            <span
              class="bp3-text-overflow-ellipsis bp3-fill"
            >
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/roles/_account_dummy"
              >
                Account dummy
              </a>
               - 
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/resource-groups/_all_resources"
              >
                All Resources
              </a>
            </span>
          </span>
          <span
            class="bp3-tag harnesstag"
          >
            <span
              class="bp3-text-overflow-ellipsis bp3-fill"
            >
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/roles/_account_viewer"
              >
                Account Viewer
              </a>
               - 
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/resource-groups/_all_resources"
              >
                All Resources
              </a>
            </span>
          </span>
          <span
            class="bp3-tag customtag"
          >
            <span
              class="bp3-text-overflow-ellipsis bp3-fill"
            >
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/roles/ufbvfbvbfv"
              >
                ufbvfbvbfv
              </a>
               - 
              <a
                class="link"
                href="/account/testAcc/home/setup/access-control/resource-groups/_all_resources"
              >
                All Resources
              </a>
            </span>
          </span>
        </div>
        <div
          class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-flex-start StyledProps--padding-top-medium"
        >
          <button
            class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main StyledProps--intent-primary Button--with-current-color Button--without-shadow"
            data-testid="addRole-UserGroup"
            type="button"
          >
            <span
              class="bp3-button-text"
            >
              common.plusNumber
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
